test/framework/ownerreference_helpers.go

```diff
@@ -0,0 +1,335 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package framework
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"sort"
+	"time"
+
+	. "github.com/onsi/gomega"
+	"github.com/pkg/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	kerrors "k8s.io/apimachinery/pkg/util/errors"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	bootstrapv1 "sigs.k8s.io/cluster-api/bootstrap/kubeadm/api/v1beta1"
+	clusterctlcluster "sigs.k8s.io/cluster-api/cmd/clusterctl/client/cluster"
+	controlplanev1 "sigs.k8s.io/cluster-api/controlplane/kubeadm/api/v1beta1"
+	addonsv1 "sigs.k8s.io/cluster-api/exp/addons/api/v1beta1"
+	"sigs.k8s.io/cluster-api/util/patch"
+)
+
+// ValidateOwnerReferencesResilience checks that expected owner references are in place, deletes them, and verifies that expect owner references are properly rebuilt.
+func ValidateOwnerReferencesResilience(ctx context.Context, proxy ClusterProxy, namespace, clusterName string, assertFuncs ...map[string]func(reference []metav1.OwnerReference) error) {
+	// Check that the ownerReferences are as expected on the first iteration.
+	AssertOwnerReferences(namespace, proxy.GetKubeconfigPath(), assertFuncs...)
+
+	clusterKey := client.ObjectKey{Namespace: namespace, Name: clusterName}
+
+	// Removes all the owner references.
+	// NOTE: we are testing the worst-case scenario, where all the owner references are removed while the
+	// reconcilers are not working; this mimics what happens with the current velero backup/restore, which is an
+	// edge case where an external system intentionally nukes all the owner references in a single operation.
+	// The assumption is that if the system can recover from this edge case, it can also handle use cases where an owner
+	// reference is deleted by mistake.
+	setClusterPause(ctx, proxy.GetClient(), clusterKey, true)
+
+	removeOwnerReferences(ctx, proxy, namespace)
+
+	setClusterPause(ctx, proxy.GetClient(), clusterKey, false)
+
+	// Annotate the clusterClass, if one is in use, to speed up reconciliation.
+	forceClusterClassReconcile(ctx, proxy.GetClient(), clusterKey)
+
+	// Check that the ownerReferences are as expected after additional reconciliations.
+	AssertOwnerReferences(namespace, proxy.GetKubeconfigPath(), assertFuncs...)
+}
+
+func AssertOwnerReferences(namespace, kubeconfigPath string, assertFuncs ...map[string]func(reference []metav1.OwnerReference) error) {
+	allAssertFuncs := map[string]func(reference []metav1.OwnerReference) error{}
+	for _, m := range assertFuncs {
+		for k, v := range m {
+			allAssertFuncs[k] = v
+		}
+	}
+	Eventually(func() error {
+		graph, err := clusterctlcluster.GetOwnerGraph(namespace, kubeconfigPath)
+		Expect(err).To(BeNil())
```

Comment:
I assume we intentionally want to fail here if we get an error once? (and not retry with eventually)

Reply:
Yeah - actually this is running retries within the func so this is unnecessary.

Reply:
Looks like the test is flaky if Eventually isn't included because of different order / speed of the reconcilers.